### PR TITLE
Fix a crash in v2i_IPAddrBlocks()

### PR DESCRIPTION
**Pull request by mattcaswell**
If an IP address prefix value is supplied that is too large then a crash
can result. v2i_IPAddrBlocks() should sanity check the prefix value, as
should X509v3_addr_add_prefix().

Reported by Theo Buehler (@botovq)

### CODE REVIEW

**Review by t8m — APPROVED** [review]

File: crypto/x509/v3_addr.c
Comment (botovq) [inline crypto/x509/v3_addr.c:593]: I would have put this check into `make_addressPrefix()`.

The API is a bit undecided on whether it wants to support a generic AFI or only IPv4 and IPv6. In LibreSSL I have opted to cap the prefixlen to `8 * 16` unless the AFI is IPv4 in which case I have capped it at `8 * 4`. Your choice is legitimate, but a bit more restrictive.

File: crypto/x509/v3_addr.c
Comment (botovq) [inline crypto/x509/v3_addr.c:596]: A prefix length of 0 is actually used by RPKI trust anchors for the special case of the entire IP address spaces `0.0.0.0/0` and `::/0`.

Here is an example: https://console.rpki-client.org/ta/ripe/ripe-ncc-ta.cer.html

File: crypto/x509/v3_addr.c
Comment (botovq) [inline crypto/x509/v3_addr.c:996]: Again, I think a prefix length of 0 should be permitted. I'm not sure it's worth it to add further checks to ensure the address is all zeroes.

File: crypto/x509/v3_addr.c
Comment (mattcaswell) [reply crypto/x509/v3_addr.c]: Yeah, ok - I've moved the check there.

File: crypto/x509/v3_addr.c
Comment (mattcaswell) [reply crypto/x509/v3_addr.c]: Ah! Right. Thanks. That makes sense. Fixed.

File: crypto/x509/v3_addr.c
Comment (mattcaswell) [reply crypto/x509/v3_addr.c]: Fixed

**Comment by mattcaswell** [comment]
Updated to address review feedback.

**Comment by mattcaswell** [comment]
@t8m - please reconfirm

**Comment by mattcaswell** [comment]
Ping for second review @openssl/committers

**Comment by mattcaswell** [comment]
Ping for second review....really want this in the release next week! @openssl/committers

File: test/v3ext.c
Comment (beldmit) [inline test/v3ext.c:85]: Shouldn't we also test windows-style EOLs?

Comment (mattcaswell) [reply test/v3ext.c:85]: I'd say no - at least in this PR. I've focused specifically on the IP address formats, rather than the extensions processing in general. This test could be expanded massively to deal with the latter - but the scope would be quite significantly more than I'm trying to deal with here.

File: test/v3ext.c
Comment (beldmit) [inline test/v3ext.c:62]: ` { "sbgp-ipAddrBlock = IPv6:2001::db8\n", 1 },` also looks like a good test - with :: in the middle.

**Review by beldmit — APPROVED** [review]
LGTM with or without extra test

**Comment by mattcaswell** [comment]
I added the additional test.

@t8m still ok?

**Comment by t8m** [comment]
There is a conflict now

**Comment by mattcaswell** [comment]
> There is a merge conflict now

I am going to wait until #18524 is merged before fixing this - since that is going to create yet another merge conflict.

**Comment by DDvO** [comment]
> > There is a merge conflict now
> 
> I am going to wait until #18524 is merged before fixing this - since that is going to create yet another merge conflict.

Looks like this meanwhile can continue.

**Comment by mattcaswell** [comment]
Rebased to resolve merge conflicts.

Ping @t8m and @beldmit. Can you reconfirm your approvals?

**Review by t8m — APPROVED** [review]

**Review by beldmit — APPROVED** [review]
Still LGTM

**Comment by mattcaswell** [comment]
Grr. There was another conflict - rebased again. This time the resolution was trivial so I will assume that the previous approvals still apply.

**Review by t8m — APPROVED** [review]

**Comment by hlandau** [comment]
@t8m Can this be merged now?

**Comment by mattcaswell** [comment]
> Can this be merged now?

Yes, it has all necessary approvals. I will merge.

**Comment by mattcaswell** [comment]
Merged to master and 3.0. There was a trivial whitespace related conflict when backporting to 3.0 that I just fixed.

There is a different conflict backporting to 1.1.1 and I will raise a separate PR for that. Closing this one.

**Pull request closed without merging**
